testsuite/cluster-test/src/atomic_histogram.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::tx_emitter::TxStats;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+const HISTOGRAM_CAPACITY: usize = 2048;
+const HISTOGRAM_WIDTH_MILLIS: u64 = 50;
+
+pub struct AtomicHistogram {
+    buckets: Vec<AtomicU64>,
+}
+
+impl AtomicHistogram {
+    pub fn default() -> Self {
+        let mut buf = Vec::with_capacity(HISTOGRAM_CAPACITY);
+        for _i in 0..HISTOGRAM_CAPACITY {
+            buf.push(AtomicU64::new(0));
+        }
+        Self { buckets: buf }
+    }
+
+    pub fn to_vec(&self) -> Vec<u64> {
+        let mut buf = Vec::with_capacity(HISTOGRAM_CAPACITY);
+        for i in 0..HISTOGRAM_CAPACITY {
+            buf.push(self.buckets[i].load(Ordering::Relaxed));
+        }
+
+        buf
+    }
+
+    fn get_bucket_num(&self, latency: u64) -> u64 {
+        let bucket_num = latency / HISTOGRAM_WIDTH_MILLIS;
+        if bucket_num >= HISTOGRAM_CAPACITY as u64 - 2 {
+            return HISTOGRAM_CAPACITY as u64 - 1;
+        }
+        bucket_num
+    }
+
+    pub fn record_latency_in_bucket(&self, latency: u64, num_requests: u64) {
+        let bucket_num = self.get_bucket_num(latency);
+        self.buckets[bucket_num as usize].fetch_add(num_requests as u64, Ordering::Relaxed);
+    }
+
+    pub fn sub(a: &[u64], b: &[u64]) -> Vec<u64> {
+        let mut buf = Vec::with_capacity(a.len());
+        if a.len() != b.len() {
+            for &i in a.iter() {
+                buf.push(i);
+            }
+        } else {
+            for i in 0..a.len() {
+                buf.push(a[i] - b[i]);
+            }
+        }
+        buf
+    }
+
+    pub fn calculate_p90(stats: &TxStats) -> u64 {
+        let p90_count = stats.committed * 9 / 10;
+        let mut counter = 0u64;
+        let mut total_latency = 0u64;
+        for i in 0..stats.latency_buckets.len() {
+            let num = stats.latency_buckets[i];
+            let gap = p90_count - counter;
+            if counter >= p90_count {
+                break;
+            }
+            if num > gap {
+                total_latency += gap * HISTOGRAM_WIDTH_MILLIS * i as u64;
+                counter += gap;
+            } else {
+                total_latency += num * HISTOGRAM_WIDTH_MILLIS * i as u64;
+                counter += num;
+            }
+        }
+        if p90_count == 0 {
+            return 0;
+        }
+        total_latency / p90_count
```

Comment:
This formula here is a bit unusual, are you sure it's correct? Is it some fancy histogram formula with improved precision?

I was thinking about something as simple as
```
for i in 0.. stats.latency_buckets.len() {
count+=stats.latency_buckets[i];
if count > p90_count { return i* HISTOGRAM_WIDTH_MILLIS}
}
```

Reply:
I think yes, I construct this formula which works by steps:

1. count p90_count
2. iterate each bucket from low to high(from 50ms to ... n * 50ms), count all txns are fitted in this bucket, sum up totally latency
3. for special case, the last bucket count is more than total p90 count, we can not just add on the whole count of that bucket, I use ```gap``` to count the last piece of p90_count

this is important because for example, with larger txn total number, we will lost more and more accuracy, for worst case we will add extra total latency = (10% * total_txn - 1) * i * HISTOGRAM_WIDTH_MILLIS

Reply:
I see. Let's cover it with unit tests then :)